xllm/core/scheduler/pd_ooc_scheduler.cpp

```diff
@@ -0,0 +1,1203 @@
+        ADD_VECTOR_TO_PROTO(
+            gen->mutable_top_logprobs(),
+            request->sequences()[0]->first_token().value().token_top_logprobs);
+      }
+      gen->set_kv_cache_transfer_mode(options_.kv_cache_transfer_mode());
+      if (options_.kv_cache_transfer_mode() == "PULL") {
+        ADD_VECTOR_TO_PROTO(gen->mutable_cluster_ids(),
+                            instance_info_.cluster_ids);
+        ADD_VECTOR_TO_PROTO(gen->mutable_addrs(), instance_info_.addrs);
+        ADD_VECTOR_TO_PROTO(gen->mutable_k_cache_ids(),
+                            instance_info_.k_cache_ids);
+        ADD_VECTOR_TO_PROTO(gen->mutable_v_cache_ids(),
+                            instance_info_.v_cache_ids);
+
+        const auto blocks = request->sequences()[0]->kv_state().kv_blocks();
+        std::vector<uint64_t> block_ids;
+        block_ids.reserve(blocks.size());
+        for (const auto& block : blocks) {
+          block_ids.push_back(block.id());
+        }
+        ADD_VECTOR_TO_PROTO(gen->mutable_block_ids(), block_ids);
+        gen->set_dp_size(instance_info_.dp_size);
+        gen->set_dp_rank(request->sequences()[0]->dp_rank());
+      }
+      //*gens.mutable_gens()->Add() = gen;
+
+      // send first gens to remote instance
+      proto::PDOOCService_Stub* stub = nullptr;
+      {
+        std::lock_guard<std::mutex> lock(req_to_channel_map_mutex_);
+        // now we only support one request once.
+        stub = req_to_channel_map_[request->request_id()];
+      }
+
+      // TODO: Async call later
+      proto::Status resp;
+      brpc::Controller cntl;
+      stub->FirstGeneration(&cntl, &gens, &resp, nullptr);
+      if (options_.enable_decode_response_to_service() || cntl.Failed() ||
+          !resp.ok()) {
+        if (cntl.Failed() || !resp.ok()) {
+          LOG(ERROR) << "Failed to send first generation, " << cntl.ErrorText()
+                     << ", staus: " << resp.ok();
+        }
+        {
+          std::lock_guard<std::mutex> lock(remote_requests_map_mutex_);
+          remote_requests_map_.erase(request->request_id());
+          remote_requests_output_thread_map_.erase(request->request_id());
+        }
+        {
+          std::lock_guard<std::mutex> lock(req_to_channel_map_mutex_);
+          req_to_channel_map_.erase(request->request_id());
+        }
+        block_manager_pool_->deallocate(request.get());
+      } else {
+        // release the memory for other requests.
+        // TODO: FIXME
+        // Here, we should decide whether to recycle the allocated blocks
+        // according to whether all the blocks have been transmitted or not.
+        block_manager_pool_->deallocate(request.get());
+      }
+    }
+  });
+}
+
+bool PDOOCScheduler::prefill_recv_generation(const RequestOutput& output) {
+  std::shared_ptr<Request> request = nullptr;
+  int request_thread_idx = -1;
+  {
+    std::lock_guard<std::mutex> lock(remote_requests_map_mutex_);
```

Comment:
It is necessary to create a concurrent map/thread_local map to encapsulate the mutex.